src/text.c

```diff
@@ -0,0 +1,441 @@
+/* sdb - MIT - Copyright 2020 - thestr4ng3r */
+
+#include "sdb.h"
+
+#include <fcntl.h>
+#include <limits.h>
+
+/**
+ * ********************
+ * Plaintext SDB Format
+ * ********************
+ *
+ * Files are UTF-8 and use '\n' line endings. Always.
+ * 
+ * Lines starting with '/' denote the path of the namespace for the following data:
+ *
+ *   /some/namespace
+ * 
+ * The default path is root, just a slash also means root.
+ * These paths are always absolute from the root. Characters that must be escaped in a path are: '/', '\\', '\n', '\r':
+ *
+ *   /s\/ome/name\nspa\\ce
+ *
+ * SDB entries are written each as a single k=v line:
+ *
+ *   somekey=somevalue
+ *
+ * To distinguish these from path lines, if there is a leading '/' in the key, it must be escaped
+ * (slashes later in the line don't have to be escaped):
+ *
+ *   \/slashedkey=somevalue
+ *
+ * Other than that, at any postion, '\\', '\n' and '\r' must be escaped:
+ *   
+ *   some\\key=some\nvalue
+ *
+ * In the key, '=' must also be escaped (not necessary in the value):
+ *
+ *   some\=key=some=value
+ * 
+ * --------
+ * Example:
+ *
+ *   /
+ *   key=intheroot
+ *   \/slashedkey=somevalue
+ *   some\\key=some\nvalue
+ *   some\=key=some=value
+ *
+ *   /subns
+ *   some=stuff in the sub-namespace
+ *
+ *   /subns/deeper
+ *   this=is in /subns/deeper
+ *
+*/
+
+static int cmp_ns(const void *a, const void *b) {
+	const SdbNs *nsa = a;
+	const SdbNs *cia = b;
+	return strcmp (nsa->name, cia->name);
+}
+
+// n = position we are currently looking at
+// p = position until we have already written everything
+// macro to flush a block of text that doesn't have to be escaped
+#define FLUSH do { if (p != n) { write (fd, p, n - p); p = n; } } while (0)
+// macro to flush and skip a char
+#define SKIP do { FLUSH; p++; } while (0)
+
+static void write_path(int fd, SdbList *path) {
+	write (fd, "/", 1); // always print a /, even if path is empty
+	SdbListIter *it;
+	const char *path_token;
+	bool first = true;
+	ls_foreach (path, it, path_token) {
+		if (first) {
+			first = false;
+		} else {
+			write (fd, "/", 1);
+		}
+		// write and escape the path
+		const char *p = path_token;
+		const char *n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
```

Comment:
this char replacing thing is done in 3 places, and its duplicating the same code for each case, when it can be simplified to just replace one char after the backslash and implemented in a single helper function. can you do this refactoring?

* write_path
* write_key
* write_value

Reply:
A function would be quite ugly here since you need to pass many pointers to local vars that have to be changed. I can do some more macro magic though to make it look nicer. Whether that is better then is a matter of taste though.

Reply:
magic applied now, please check again